### PR TITLE
AWSRecreateSG Improvements

### PR DESCRIPTION
**Pull request by johnnywilkes**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
https://jira-dc.paloaltonetworks.com/browse/EXPANDR-5777

## Description
AWSRecreateSG
- Added the **assume_role** optional input in case an roleArn needs to be passed in.
- Updated the script to work even when there are multiple AWS-EC2 integration instances configured.
## Must have
- [X] Tests
- [X] Documentation

### CODE REVIEW

**Comment by johnnywilkes** [comment]
@ShirleyDenkberg, can you please review?

File: Packs/AWS-Enrichment-Remediation/ReleaseNotes/1_1_4.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
+
+#### Scripts
+
+##### AWSRecreateSG
+
+- Added the **assume_role** optional input in case an roleArn needs to be passed in.
```
Comment (ShirleyDenkberg) [inline Packs/AWS-Enrichment-Remediation/ReleaseNotes/1_1_4.md:6]:
```suggestion
- Added the **assume_role** optional input in case a roleArn needs to be passed in.
```

**Comment by ShirleyDenkberg** [comment]
@MichaelYochpaz Doc review completed.

File: Packs/AWS-Enrichment-Remediation/Scripts/AWSRecreateSG/AWSRecreateSG.py [hunk 2]
```diff
@@ -208,30 +236,37 @@ def instance_info(instance_id: str, public_ip: str) -> Dict:
...
                     group_list = []
                     for sg in interface['Groups']:
                         group_list.append(sg['GroupId'])
                     mapping_dict[interface['NetworkInterfaceId']] = group_list
-        if match is False:
-            raise ValueError('could not find interface with public IP association')
-        return mapping_dict
-    else:
-        raise ValueError('failed to pull information on EC2 instance')
+                    integration_to_use = instance['ModuleName']
```
Comment (MichaelYochpaz) [inline Packs/AWS-Enrichment-Remediation/Scripts/AWSRecreateSG/AWSRecreateSG.py:263]: Let's add a `break` here. No need to continue looping if we found our instance.
(may have messed up the correct indentation in this suggestion, probably better to add this line in an IDE)
```suggestion
                    integration_to_use = instance['ModuleName']
            break
```

File: Packs/AWS-Enrichment-Remediation/Scripts/AWSRecreateSG/AWSRecreateSG.py [hunk 3]
```diff
@@ -110,14 +113,21 @@ def sg_fix(sg_info: List, port: int, protocol: str) -> Dict:
             recreate_list.append(str(priv).replace("'", "\""))
         new_name = info['GroupName'] + "_xpanse_ar_" + str(randint(100, 999))
         description = "copied from rule " + info['GroupName'] + " by Xpanse Active Response module"
-        new_sg = demisto.executeCommand("aws-ec2-create-security-group",
-                                        {"groupName": new_name, "vpcId": info['VpcId'], "description": description})
+        cmd_args = {"groupName": new_name, "vpcId": info['VpcId'], "description": description, "using": integration_to_use}
+        if assume_role:
+            cmd_args.update({'roleArn': assume_role})
+            cmd_args.update({'roleSessionName': "test"})
```
Comment (MichaelYochpaz) [inline Packs/AWS-Enrichment-Remediation/Scripts/AWSRecreateSG/AWSRecreateSG.py:119]: So, the value of `roleSessionName` is going to always be`test`?
Is that not a confusing name?
Also, wouldn't that cause a conflict if there are multiple sessions running at the same time, and they're all named `test`?

(this note is true for other places in the PR where the `roleSessionName` is used)

File: Packs/AWS-Enrichment-Remediation/Scripts/AWSRecreateSG/AWSRecreateSG.yml [hunk 4]
```diff
@@ -39,3 +41,4 @@ type: python
 fromversion: 6.5.0
 tests:
 - No tests (auto formatted)
+runonce: false
```
Comment (MichaelYochpaz) [inline Packs/AWS-Enrichment-Remediation/Scripts/AWSRecreateSG/AWSRecreateSG.yml:44]: Could you elaborate on why this change is needed?

File: Packs/AWS-Enrichment-Remediation/Scripts/AWSRecreateSG/AWSRecreateSG.yml [hunk 5]
```diff
@@ -15,6 +15,8 @@ args:
 - description: Public IP address of the EC2 instance.
   name: public_ip
   required: true
+- description: If assuming roles for AWS, this is the name of the role to assume (should be the same for all organizations).
```
Comment (MichaelYochpaz) [inline Packs/AWS-Enrichment-Remediation/Scripts/AWSRecreateSG/AWSRecreateSG.yml:18]: Wouldn't something like

```suggestion
- description: Name of an AWS role to assume (should be the same for all organizations).
```
be more concise and in-line with the other descriptions?

(If you do change that, make sure to also change in the README)

File: Packs/AWS-Enrichment-Remediation/ReleaseNotes/1_1_4.md [hunk 6]
```diff
@@ -0,0 +1,8 @@
+
+#### Scripts
+
+##### AWSRecreateSG
+
+- Added the **assume_role** optional input in case a roleArn needs to be passed in.
+- Updated the script to work even when there are multiple AWS-EC2 integration instances configured.
```
Comment (MichaelYochpaz) [inline Packs/AWS-Enrichment-Remediation/ReleaseNotes/1_1_4.md:7]: Think this might be a bit better
```suggestion
- Updated the script to work with multiple AWS-EC2 integration instances configured.
```

**Review by MichaelYochpaz — COMMENTED** [review]
Hey, thank you for the contribution. Great job!

Wrote a few notes, but a lot of them are for the same thing repeating in multiple places, and all of them are minor changes.

Looks good overall :)

File: Packs/AWS-Enrichment-Remediation/Scripts/AWSRecreateSG/AWSRecreateSG.yml (hunk 5 above)
Comment (johnnywilkes) [reply Packs/AWS-Enrichment-Remediation/Scripts/AWSRecreateSG/AWSRecreateSG.yml:18]: i don't think I want to change because I want to note that it is actually optional

File: Packs/AWS-Enrichment-Remediation/ReleaseNotes/1_1_4.md (hunk 6 above)
Comment (johnnywilkes) [reply Packs/AWS-Enrichment-Remediation/ReleaseNotes/1_1_4.md:7]: Shirley already reviewed so I prefer to keep as-is

File: Packs/AWS-Enrichment-Remediation/Scripts/AWSRecreateSG/AWSRecreateSG.py (hunk 3 above)
Comment (johnnywilkes) [reply Packs/AWS-Enrichment-Remediation/Scripts/AWSRecreateSG/AWSRecreateSG.py:119]: I have tested and it wasn't an issue, I will name it something else that should be a bit better

File: Packs/AWS-Enrichment-Remediation/Scripts/AWSRecreateSG/AWSRecreateSG.yml (hunk 4 above)
Comment (johnnywilkes) [reply Packs/AWS-Enrichment-Remediation/Scripts/AWSRecreateSG/AWSRecreateSG.yml:44]: not sure how/why that was added.  Will remove

Comment (johnnywilkes) [reply Packs/AWS-Enrichment-Remediation/Scripts/AWSRecreateSG/AWSRecreateSG.yml]: change in [review updates](https://github.com/demisto/content/pull/29311/commits/d0b01a94de4508529522275932721c124c094d62)

File: Packs/AWS-Enrichment-Remediation/Scripts/AWSRecreateSG/AWSRecreateSG.py (hunk 3 above)
Comment (johnnywilkes) [reply Packs/AWS-Enrichment-Remediation/Scripts/AWSRecreateSG/AWSRecreateSG.py]: change in [review updates](https://github.com/demisto/content/pull/29311/commits/d0b01a94de4508529522275932721c124c094d62)

File: Packs/AWS-Enrichment-Remediation/Scripts/AWSRecreateSG/AWSRecreateSG.py (hunk 2 above)
Comment (johnnywilkes) [reply Packs/AWS-Enrichment-Remediation/Scripts/AWSRecreateSG/AWSRecreateSG.py]: change in [review updates](https://github.com/demisto/content/pull/29311/commits/d0b01a94de4508529522275932721c124c094d62)

File: Packs/AWS-Enrichment-Remediation/Scripts/AWSRecreateSG/AWSRecreateSG.py [hunk 7]
```diff
@@ -50,10 +52,10 @@ def split_rule(rule: Dict, port: int, protocol: str) -> List[Dict]:
                          'UserIdGroupPairs': [], 'FromPort': port + 1, 'ToPort': 65535},
                         {'IpProtocol': 'tcp', 'IpRanges': [{'CidrIp': '0.0.0.0/0'}], 'Ipv6Ranges': [], 'PrefixListIds': [],
                          'UserIdGroupPairs': [], 'FromPort': 0, 'ToPort': 65535}]
-    return(res_list)
+    return (res_list)
 
 
-def sg_fix(sg_info: List, port: int, protocol: str) -> Dict:
+def sg_fix(sg_info: list, port: int, protocol: str, assume_role: str, integration_to_use: str) -> dict:
```
[the remark this answers is not in the document]
Comment (johnnywilkes) [reply Packs/AWS-Enrichment-Remediation/Scripts/AWSRecreateSG/AWSRecreateSG.py]: change in [review updates](https://github.com/demisto/content/pull/29311/commits/d0b01a94de4508529522275932721c124c094d62)

File: Packs/AWS-Enrichment-Remediation/Scripts/AWSRecreateSG/AWSRecreateSG.py [hunk 8]
```diff
@@ -249,17 +284,18 @@ def aws_recreate_sg(args: Dict[str, Any]) -> str:
     port = int(args.get('port', None))
     protocol = args.get('protocol', None)
     public_ip = args.get('public_ip', None)
+    assume_role = args.get('assume_role', None)
```
[the remark this answers is not in the document]
Comment (johnnywilkes) [reply Packs/AWS-Enrichment-Remediation/Scripts/AWSRecreateSG/AWSRecreateSG.py:280]: FYI, i readded these because it was causing MyPy errors and don't seem to hurt anything

**Comment by johnnywilkes** [comment]
@MichaelYochpaz , changes are complete, please let me know if you need anything else

File: Packs/AWS-Enrichment-Remediation/Scripts/AWSRecreateSG/AWSRecreateSG.py (hunk 2 above)
Comment (MichaelYochpaz) [reply Packs/AWS-Enrichment-Remediation/Scripts/AWSRecreateSG/AWSRecreateSG.py]: This will only break from the internal interfaces loop, but not the`for instance in instance_info` loop (it will keep searching the other instances even when a match is already found)

Something like this should work (my previous suggestion wasn incorrect):
```python
    for instance in instance_info:
        # Check if returned error, in the case of multiple integration instances only one should pass.
        if not isError(instance) and \
           instance.get('Contents').get('AWS.EC2.Instances(val.InstanceId === obj.InstanceId)')[0].get('NetworkInterfaces'):
            interfaces = instance.get('Contents').get(
                'AWS.EC2.Instances(val.InstanceId === obj.InstanceId)')[0].get('NetworkInterfaces')
            mapping_dict = {}
            for interface in interfaces:
                if interface.get('Association') and interface.get('Association').get('PublicIp') == public_ip:
                    match = True
                    group_list = []
                    for sg in interface['Groups']:
                        group_list.append(sg['GroupId'])
                    mapping_dict[interface['NetworkInterfaceId']] = group_list
                    instance_to_use = instance['ModuleName']
                    break  # <- What's already added

        if match:  # <- I've added this
            break

    if match is False:
        raise ValueError('could not find interface with public IP association')

```

File: Packs/AWS-Enrichment-Remediation/Scripts/AWSRecreateSG/AWSRecreateSG.yml (hunk 5 above)
Comment (MichaelYochpaz) [reply Packs/AWS-Enrichment-Remediation/Scripts/AWSRecreateSG/AWSRecreateSG.yml:18]: We don't specify in the descriptions that parameters that they are optional (unless there's some specific reason it's needed), we use the `required` field for that.

You can check descriptions for optional parameters in commands to see examples.

File: Packs/AWS-Enrichment-Remediation/Scripts/AWSRecreateSG/AWSRecreateSG.py (hunk 3 above)
Comment (MichaelYochpaz) [reply Packs/AWS-Enrichment-Remediation/Scripts/AWSRecreateSG/AWSRecreateSG.py]: Let's set a constant at the top, something like `ROLE_SESSION_NAME = "xsoar-session"`, and use that as the value.
This way we ever decide to change that for some reason, we change it in one place instead of multiple places.

File: Packs/AWS-Enrichment-Remediation/Scripts/AWSRecreateSG/AWSRecreateSG.py (hunk 2 above)
Comment (johnnywilkes) [reply Packs/AWS-Enrichment-Remediation/Scripts/AWSRecreateSG/AWSRecreateSG.py]: change made in [update 9/7](https://github.com/demisto/content/pull/29311/commits/cea008683bf63abf4ae2c6a3f1ece7795b5b5e46)

**Comment by johnnywilkes** [comment]
@MichaelYochpaz , changes are complete, please let me know if you need anything else

**Comment by johnnywilkes** [comment]
@MichaelYochpaz , I added a small change to the playbook that references that automation

**Review by MichaelYochpaz — APPROVED** [review]
Looks good :)
Good job!
